R/conditions.R

```diff
@@ -161,3 +176,129 @@ skip_internal_condition <- function(cnd) {
     cnd
   }
 }
+
+
+# Warnings -------------------------------------------------------------
+
+#' Show warnings from the last command
+#'
+#' Warnings that occur inside a dplyr verb like `mutate()` are caught
+#' and stashed away instead of being emitted to the console. This
+#' prevents rowwise and grouped data frames from flooding the console
+#' with warnings. To see the original warnings, use
+#' `last_dplyr_warnings()`.
+#'
+#' @param n Passed to [head()] so that only the first `n` warnings are
+#'   displayed.
+#' @export
+last_dplyr_warnings <- function(n = 5) {
+  if (!identical(n, Inf)) {
+    check_number(n)
+    stopifnot(n >= 0)
+  }
+
+  warnings <- the$last_warnings
+  remaining <- max(length(warnings) - n, 0L)
+
+  warnings <- head(warnings, n = n)
+  warnings <- map(warnings, new_dplyr_warning)
+
+  structure(
+    warnings,
+    class = c("last_dplyr_warnings", "list"),
+    remaining = remaining
+  )
+}
+
+the$last_warnings <- list()
+the$last_cmd_frame <- ""
+
+# Flushes warnings if a new top-level command is detected
+push_dplyr_warnings <- function(warnings) {
+  last <- the$last_cmd_frame
+  current <- obj_address(sys.frame(1))
+
+  if (!identical(last, current)) {
+    reset_dplyr_warnings()
+    the$last_cmd_frame <- current
+  }
+
+  the$last_warnings <- c(the$last_warnings, warnings)
+}
+
+# Also used in tests
+reset_dplyr_warnings <- function() {
+  the$last_warnings <- list()
+}
+
+signal_warnings <- function(warnings, error_call) {
+  n <- length(warnings)
+  if (!n) {
+    return()
+  }
+
+  push_dplyr_warnings(warnings)
+
+  first <- warnings[[1]][["cnd"]]
+  call <- format_error_call(error_call)
+
+  cli::cli_warn(c(
+    "There {cli::qty(n)} {?was/were} {n} warning{?s} in a {call} step.",
+    if (n > 1) "The first warning was:",
+    cnd_message_lines(first),
+    i = if (n > 1) "Run `dplyr::last_dplyr_warnings()` to see the {n - 1} remaining warning{?s}."
```

Comment:
```suggestion
    i = if (n > 1) "Run `dplyr::last_dplyr_warnings()` to see the {n - 1} remaining warning{?s}."
```

I WANT TO BE ABLE TO CLICK THIS `dplyr::last_dplyr_warnings()`

Reply:
Maybe we can wait for Gabor to finish implementing link syntax in the next cli version?

Reply:
Yeah I think we could do it in one sweep.